### ISSUE TITLE
Callable type is not properly inferred

### ISSUE THREAD

**Issue by Jonxslays** [microsoft/pyright#3386]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Pyright produces error on strict when using `typing.Callable` without any generic arguments.

**To Reproduce**
Steps to reproduce the behavior.

```toml
[tool.pyright]
typeCheckingMode = "strict"
reportMissingTypeArgument = false
```

```py
import typing

def func(func_arg: typing.Callable) -> typing.Callable:
    return func_arg

f = func
```

`func_arg`, `func` and `f` are all unknown in this scenario.

**Expected behavior**
According to the typing documentation [here](https://docs.python.org/3/library/typing.html#typing.Callable):

> A plain [Callable](https://docs.python.org/3/library/typing.html#typing.Callable) is equivalent to Callable[..., Any], and in turn to [collections.abc.Callable](https://docs.python.org/3/library/collections.abc.html#collections.abc.Callable).

Based on this it seems like a `Callable` without the generics should be interpreted as `Callable[..., Any]` rather than reporting unknown.

**VS Code extension or command-line**
Extension version: 1.1.239
CLI version: 1.1.235

**Comment by erictraut** [comment]
You have omitted the type arguments for `Callable`. Pyright therefore infers `Unknown` for these type arguments. `Unknown` is a special form of `Any` that pyright uses for an implicit `Any`. It is a common programming error to omit type arguments unintentionally and end up with an `Any` when that is not desired. This masks type violations. 

If you want to use "strict" type checking, you need to provide explicit type arguments for `Callable` and other generic types.

If you don't care about strict type checking, you can enable only a subset of diagnostic checks.

**Comment by davidism** [comment]
But that's not correct according to Python's documentation, which was linked and quoted.

**Comment by erictraut** [comment]
It is correct and consistent with PEP 484. Pyright treats an `Unknown` as the equivalent of `Any` when performing type compatibility checks. You'll note that your sample doesn't generate any diagnostics related to type compatibility. The annotation `Callable` is treated as the equivalent of `Callable[..., Any]`. It's just that pyright notes that the `Any` in this case was implicit, not explicit.

**Comment by davidism** [comment]
OK, so pyright strict mode means "no assumptions about implicit arguments, even if they are documented"?

**Comment by davidism** [comment]
That's basically opening up every open source project that adds typing to issues from users who don't understand that pyright strict mode doesn't follow the documented typing behavior. A project can add typing that is correct according to the documentation, but still be reported as incorrect when users run pyright.

**Comment by erictraut** [comment]
Yes, that's correct. Implicit type arguments are a common source of bugs, so pyright offers the ability to flag cases where `Unknown` (i.e. implicit `Any`) is used in your code. This gives you a chance to fix the potential hole in type validation. You still have the option of using `Any` explicitly (as in `list[Any]` or `Callable[..., Any]`), but in most cases it's better to provide type arguments that are not `Any`].

You are correct that python's strict mode requires library authors to provide accurate and complete types. The use of `Any` type arguments (implicit or otherwise) is not recommended for libraries because it leads to holes in type checking and produces sub-par experiences for consumers of the library.

**Comment by davidism** [comment]
But `Callable` on its own is accurate and complete according to Python's typing documentation.

**Comment by parafoxia** [comment]
It may be prudent to include a note whenever this error appears just briefly explaining this, perhaps with a link to a more detailed explanation if the user wanted to find out more. It would prevent further confusion on the matter.

**Comment by davidism** [comment]
Here's the relevant code in Click: https://github.com/pallets/click/blob/6bd8e56de957a423a707362cecfe2d5560afba0b/src/click/decorators.py#L127-L155 A Click command can return an arbitrary value, to support pipelines that collect and process chained command results.

**Comment by davidism** [comment]
I'm about to attend PyCon, and will be running two day of sprints. It sounds like a good task would be to run pyright and fix this completeness. Is there a specific command beyond `pyright --verifytypes click`  that contributors should use? If you or other pyright people are at PyCon, I'd be happy to meet up and chat or have an open space.

I don't have the reference Python issue/PR, but inheriting from `Generic` at runtime is significantly slower until maybe Python 3.9. The other option is to use a `pyi` file, but my experience has been that it is very hard to continually maintain two files in sync.

**Comment by erictraut** [comment]
Yeah, just install the latest version of the library and then run `pyright --verifytypes click`. It will emit diagnostics for each symbol that is either partially unknown or "ambiguous" (i.e. its type can be inferred by pyright but could be inferred differently by other type checkers). If you want even more detailed information, such as a full list of all symbols, you can use `pyright --verifytypes click --verbose`. You can also ask pyright to emit the output in JSON form, which is useful for CI scripts or as part of other toolchains (`pyright --verifytypes click --outputjson`).

Let me know if you run into any problems or have questions or feedback.

I won't be at PyCon this year, so I won't be able to meet there.
